src/Cardano/Wallet/Mnemonic.hs

```diff
@@ -0,0 +1,332 @@
+    { mnemonicToEntropy  :: Entropy (EntropySize mw)
+    , mnemonicToSentence :: MnemonicSentence mw
+    } deriving (Eq, Show)
+
+
+newtype MnemonicException csz =
+    UnexpectedEntropyError (EntropyError csz)
+    deriving (Show, Typeable)
+
+
+data MnemonicError csz
+    = ErrMnemonicWords MnemonicWordsError
+    | ErrEntropy (EntropyError csz)
+    | ErrDictionary DictionaryError
+    | ErrForbidden
+    deriving (Show)
+
+
+
+-- | Smart-constructor for the Entropy
+mkEntropy
+    :: forall n csz. (ValidEntropySize n, ValidChecksumSize n csz)
+    => ByteString
+    -> Either (EntropyError csz) (Entropy n)
+mkEntropy = toEntropy
+
+
+ambiguousNatVal
+    :: forall n . (KnownNat n)
+    => Integer
+ambiguousNatVal = natVal @n Proxy
+
+-- | Generate Entropy of a given size using a random seed.
+--
+-- Example:
+--     do
+--       ent <- genEntropy :: IO (Entropy 12)
+genEntropy
+    :: forall n csz. (ValidEntropySize n, ValidChecksumSize n csz)
+    => IO (Entropy n)
+genEntropy =
+    let
+        size =
+            fromIntegral $ ambiguousNatVal @n
+        eitherToIO =
+            either (throwM . UnexpectedEntropyError) return
+    in
+        (eitherToIO . mkEntropy) =<< Crypto.getEntropy (size `div` 8)
+
+
+-- | Smart-constructor for the Mnemonic
+mkMnemonic
+    :: forall mw n csz.
+     ( ConsistentEntropy n mw csz
+     , EntropySize mw ~ n
+     )
+    => [Text]
+    -> Either (MnemonicError csz) (Mnemonic mw)
+mkMnemonic wordsm = do
+    phrase <- left ErrMnemonicWords
+        $ mnemonicPhrase @mw (toUtf8String <$> wordsm)
+
+    sentence <- left ErrDictionary
+        $ mnemonicPhraseToMnemonicSentence Dictionary.english phrase
+
+    entropy <- left ErrEntropy
+        $ wordsToEntropy sentence
+
+    when (isForbiddenMnemonic sentence) $ Left ErrForbidden
+
+    pure Mnemonic
+        { mnemonicToEntropy  = entropy
+        , mnemonicToSentence = sentence
+        }
+
+
+-- | Convert an Entropy to a corresponding Mnemonic Sentence
+entropyToMnemonic
+    :: forall mw n csz.
+     ( ValidMnemonicSentence mw
+     , ValidEntropySize n
+     , ValidChecksumSize n csz
+     , n ~ EntropySize mw
+     , mw ~ MnemonicWords n
+     )
+    => Entropy n
+    -> Mnemonic mw
+entropyToMnemonic entropy = Mnemonic
+    { mnemonicToSentence = entropyToWords entropy
+    , mnemonicToEntropy  = entropy
+    }
+
+
+-- | Convert 'Entropy' to a raw 'ByteString'
+entropyToByteString
+    :: Entropy n
+    -> ByteString
+entropyToByteString = entropyRaw
+
+
+-- | Constant-time comparison of any sentence with the 12-word example mnemonic
+isForbiddenMnemonic
+    :: (ValidMnemonicSentence mw)
+    => MnemonicSentence mw
+    -> Bool
+isForbiddenMnemonic sentence =
+    let
+        bytes =
+            sentenceToRawString sentence
+
+        forbiddenMnemonics = sentenceToRawString <$>
+            [ mnemonicToSentence (def @(Mnemonic 12))
+            ]
+    in
+        any (constEq bytes) forbiddenMnemonics
```

Comment:
Let's also get rid of this one and make sure (as it's now in the spec) that the mnemonics from the documentation represent invalid mnemonics (so that we don't have to have an extra hard-coded mnemonic here..

Reply:
in the spec we have (Mnemonic 15) - you mean - put the exact example from the spec and correct the affected tests, right?

Reply:
I mean, not check for forbidden mnemonics. They won't be needed if the mnemonic we put in the documentation is invalid (so by essence, it can't be used) such that we can remove all this logic from our way.

Reply:
done, added respective test to make sure mnemonic example from API is invalid:
```
114         it "Mnemonic from Api is invalid" $ do                                                                                                                                            
115             let mnemonicFromApi =                                                                                                                                                         
116                     "[squirrel,material,silly,twice,direct,slush,pistol,razor,become,junk,kingdom,flee,squirrel,silly,twice]"                                                             
117             (mkMnemonic @15 . extractWords) mnemonicFromApi `shouldSatisfy` isLeft
```

Reply:
In detail it gives:
```
Left (ErrEntropy (ErrInvalidEntropyChecksum (Checksum 25) (Checksum 26)))
```